hudi-common/src/main/java/org/apache/hudi/common/model/HoodiePartitionMetadata.java

```diff
@@ -92,11 +92,12 @@ public int getPartitionDepth() {
 
   /**
    * Write the metadata safely into partition atomically.
+   * To avoid concurrent write into the same partition (for example in speculative case),
+   * please make sure writeToken is unique.
    */
-  public void trySave(int taskPartitionId) {
+  public void trySave(String writeToken) throws IOException {
```

Comment:
Can it be a UUID, why we need to pass around the writeToken ?

Reply:
Here we want a unique suffix for temp metadata file, and `writeToken` just fullfills.

Reply:
Maybe we can utililize thee method `HoodieWrapperFileSystem#createImmutableFileInPath` and make the suffix a UUID.

Reply:
It's good to reuse such write file code, but `fs` is not `HoodieWrapperFileSystem` in this class, it's not easy to reuse `createImmutableFileInPath()` method here, do we need move this method into `FSUtils`?
https://github.com/apache/hudi/blob/7c55ac35ba11ed00151bf9d536aecdb5d83af33f/hudi-common/src/main/java/org/apache/hudi/common/model/HoodiePartitionMetadata.java#L61

As for writeToken and UUID, we prefer writeToken because it includes the task partition info which is useful in debug, pls let me know your idea.

Reply:
Both UUID and writeToken could make the metadata temp file name is unique for different task.
+1 on writeToken because it contains more information to help us find the tasks.

Reply:
>  it's not easy to reuse createImmutableFileInPath() method here

Needs some refactoring to the method, we can move it to FSUtils as a static method and make the suffix configurable.

Reply:
hi everyone, I has cherry pick this patch to our internal, then if we enable parallel write, there is a high probability that `tmpMetaPath` write will fail because it was deleted by another thread.
Maybe UUID is a good choice.

Reply:
Hey, @Tartarus0zm what is parallel write? 2 jobs writing to the same path?

Reply:
> Hey, @Tartarus0zm what is parallel write? 2 jobs writing to the same path?

Sorry, I need to explain here.
`parallel write` is our internal optimization strategy that uses multi-threading to write out hdfs files to reduce the time consuming flink Checkpoint.

Reply:
Let's switch to UUID to simplify the code, there is no possibility for conflicts any more, so the gains for debugging is little, we just need to handle the tmp file deletion in good manner, which should be ensured by `HoodieWrapperFileSystem#createImmutableFileInPath`.

Reply:
I see, we don't need UUID, after using `RetryHelper`, this issue should be fixed:

```java
metafileExists = fs.exists(metaPath);
      if (!metafileExists) {
        // write to temporary file
        writeMetafile(tmpMetaPath);
        // move to actual path
        fs.rename(tmpMetaPath, metaPath);
      }
```
If 2 threads concurrently creating tmpMetaPath, after one thread renamed succeed, and another one failed because of tmpMetaPath missing, it will retry and return directly since `metafileExists` becomes true

Reply:
Yeah, the `HoodieWrapperFileSystem#createImmutableFileInPath` can actually ensure atomicity.

Reply:
@danny0405 @boneanxs  Using `RetryHelper` means using `HoodieRetryWrapperFileSystem` directly in HoodiePartitionMetadata ?

Reply:
` RetryHelper` should be suitable to use here.

Reply:
So sorry that I have pushed the patch to another repository and did not realize it :'(. 
I have added the retry logic, please help take a look again. @beyond1920  @boneanxs @danny0405 @Tartarus0zm

Reply:
I have fied a fix for the random suffix, https://github.com/apache/hudi/pull/11052